### ISSUE TITLE
Better assets integration

### ISSUE THREAD

**Issue by andrescevp** [symfony/symfony#15320]
When i try to add some JS library, like require.js I must install a specific bundle with a lot files like this repo:

https://github.com/hearsayit/HearsayRequireJSBundle

I think that all js assets must be more transparent, other example I try to add Ace editor and the problem was when I was try add a extension for this editor, the JS files not index correctly between them.

**Comment by fabpot** [comment]
What are you suggesting here?

**Comment by andrescevp** [comment]
I mean that the assets system should be more flexible, it's just a suggestion for nexts versions.

Currently add complex JS sometimes it's really complex and not friendly.

**Comment by xabbuh** [comment]
@andrescevp Yeah, but what exactly would make it easier for you?

**Comment by andrescevp** [comment]
Will be cool if we can add assets without content rewrite, for example if you want add BootStrap (twitter) you must specify the routes like: 'bundles/some/..." if you try access it with "@AppController/..." the content is rewrited and you miss the glyphicons config. Or if you want add some extra JS and it have some dependency... my last experiment was with ace code editor and I try add some extension for it the routes are completly wrong and then the main JS don't found sorrectly the route for the extension JS.

I don't know if currently you have some way for make it, but for this reason i think that assets should be more transparent.

Thanks.

**Comment by stof** [comment]
@andrescevp when talking about content rewrite, this makes me think you are referring to Assetic. Assetic is not mandatory at all to include assets. If you don't want the assetic features, build the assets the way you want (using grunt or gulp as build tools for instance, or writing directly assets in the format used by the frontend), and then just use the Symfony asset system to reference them

**Comment by andrescevp** [comment]
Okey thanks very much for your comments, any way will be nice if someone can review this and give better native support in the framework for not need other tools.

**Comment by jakzal** [comment]
@andrescevp I still fail to see what you'd like to improve and where (Symfony or [Assetic](https://github.com/symfony/AsseticBundle)?). We can't do much if all you request is "Better native support for JS". You'll need to be more precise.

**Comment by andrescevp** [comment]
Yes I know, how I told before It is not something concrete. It is rather something general. I put a very clear example before I think. For example if you want add require.js you need make some stuff before to can use it. Fortunately Symfony have a really big community and you can found some bundle just for it. But it should be more easy, just put your assets and run it.

Other example where I'm trying make some cool stuff s with ace code editor, and for get the extensions you need specify a route not a name but symfony make some stuff and broken these routes and the direct integration.

well i hope that you can understand me. Thanks and greetings. :)

**Comment by stof** [comment]
@andrescevp you need to do additional stuff if you want to integrate require.js into **Assetic**. and this is logical as Assetic is the asset build system here, and so it needs to be aware of require.js.
But Symfony does not enforce the usage of Assetic. It is perfectly possible to use a gulp-based build system and then just reference the output of this build with the Symfony Asset component.
And note that in this case, you will still need to make some configuration to integrate require.js, but it will be in your gulp config rather in the Assetic config.

**Comment by fvsch** [comment]
@andrescevp If I understand your request correctly, what is not working well for you *with Assetic* is using JS libraries that need to dynamically request other files (images, JS files etc.) from a given folder.

For instance if you have a UI lib looking like this:

```
cool-lib/
    cool-lib.min.js
    i18n/
        es.json
        pt.json
    images/
        …
```

If you put this lib somewhere in a bundle and reference `.../cool-lib/cool-lib.min.js` with Assetic (using a `{% javascripts %}` tag in Twig, for instance), Assetic will copy this JS file but not the full folder. And you need the full folder to be publicly accessible, because `cool-lib.min.js` will dynamically insert DOM nodes that use some of the lib's images, it will do XHR requests to get the right JSON translation files, that kind of thing.

The simple solution is to just make the lib's directory public. You can put it directly in the `web` directory and make sure your `.htaccess` allows browsers to access it. Or you can put it in the `Resources/public` folder of your bundle, install the bundle assets with `php app/console assets:install --symlink` (if you don't already have the right symlink in `web/bundles`, and reference the JS file directly from your templates:

```twig
<script src="{{ asset('bundles/myexcellentbundle/cool-lib/cool-lib.min.js') }}"></script>
```

Basically, you don't have to use Assetic, and Symfony offers rather simple ways to expose public assets (the `Resources/public` folder for bundles, the `web` folder itself otherwise).

Also, note that the `asset()` function above mostly tweaks the URL you give it, adding a cache-busting query string at the end if you configured it. It doesn't do magic stuff like locating the right file in your project or copying files from one place to another.

**Comment by xabbuh** [comment]
Closing here as the issue is not concrete enough. Feel free to provide some concrete ideas about what could be done in the component to increase the user experience and we can reopen here.
